server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java

```diff
@@ -867,33 +829,91 @@ private RepositoryData safeRepositoryData(long repositoryStateId, Map<String, Bl
+                }
+            });
+        }
+    }
+
+    /**
+     * The result of removing a snapshot from a shard folder in the repository.
+     *
+     * @param indexId       Index that the snapshot was removed from
+     * @param shardId       Shard id that the snapshot was removed from
+     * @param newGeneration Id of the new index-${uuid} blob that does not include the snapshot any more
+     * @param blobsToDelete Blob names in the shard directory that have become unreferenced in the new shard generation
+     */
+    private record ShardSnapshotMetaDeleteResult(
+        IndexId indexId,
+        int shardId,
+        ShardGeneration newGeneration,
+        Collection<String> blobsToDelete
+    ) {}
+
+    // ---------------------------------------------------------------------------------------------------------------------------------
+    // The overall flow of execution
+
     /**
      * After updating the {@link RepositoryData} each of the shards directories is individually first moved to the next shard generation
      * and then has all now unreferenced blobs in it deleted.
      *
      * @param snapshotIds       SnapshotIds to delete
-     * @param repositoryStateId Expected repository state id
-     * @param foundIndices      All indices folders found in the repository before executing any writes to the repository during this
-     *                          delete operation
-     * @param rootBlobs         All blobs found at the root of the repository before executing any writes to the repository during this
-     *                          delete operation
-     * @param repositoryData    RepositoryData found the in the repository before executing this delete
+     * @param originalRepositoryDataGeneration {@link RepositoryData} generation at the start of the process.
+     * @param originalIndexContainers          All index containers at the start of the operation, obtained by listing the repository
+     *                                         contents.
+     * @param originalRootBlobs                All blobs found at the root of the repository at the start of the operation, obtained by
+     *                                         listing the repository contents.
+     * @param originalRepositoryData           {@link RepositoryData} at the start of the operation.
+     * @param repositoryFormatIndexVersion     The minimum {@link IndexVersion} of the nodes in the cluster and the snapshots remaining in
+     *                                         the repository.
      * @param listener          Listener to invoke once finished
      */
     private void doDeleteShardSnapshots(
         Collection<SnapshotId> snapshotIds,
-        long repositoryStateId,
-        Map<String, BlobContainer> foundIndices,
-        Map<String, BlobMetadata> rootBlobs,
-        RepositoryData repositoryData,
-        IndexVersion repoMetaVersion,
+        long originalRepositoryDataGeneration,
```

Comment:
Why do we need this parameter here when the other `RepositoryData` parameter already contains this information?

Reply:
Eventually this will be gone too I think